### PR TITLE
fix(pubsub): treat some kUnavailable as success

### LINKED ISSUE #5506: Subscriber::Subscribe can fail with kUnavailable
I do not have a reliable repro at this time,  anecdotally, it seems that if there are no messages in the subscription the service may return:

```
The service was unable to fulfill your request. Please try again. [code=8a75] [UNAVAILABLE]
```

Though it takes over a minute for this error code to be generated. Seems like we should treat that as a successful connection and restart the whole reconnection loop (as opposed to an unsuccessful connection).

### PR DESCRIPTION
**Pull request by coryan**
The service closes streaming pull requests with `kUnavailable` if there
is no data to pull. We should not treat these as failures (not even as
transient failures), because we showed that we successfully connected to
the service, that the subscription exists, and that we have permissions
to pull data. The "problem" is that we do not have data yet.

Fixes #5506

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5510)

### CODE REVIEW

**Review by devbww — APPROVED** [review]

**Review by devbww — COMMENTED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5510)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @coryan)

**Review by mr-salty — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5510#-:-MMJGotK--fQYfd2xyye:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @coryan)

___
*[google/cloud/pubsub/internal/streaming_subscription_batch_source.cc, line 201 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5510#-MMJEFwy0RuvK1HsO9tj:-MMJEFwy0RuvK1HsO9tk:b-awr8nj) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4b5e820c249ec1744ae1b9a38bcee284160f13a6/google/cloud/pubsub/internal/streaming_subscription_batch_source.cc#L201)):*
> ```
>  is a retryable failure, and
> ```

I'd strike "is a retryable failure, and", it confused me - "This should be treated..." seems clear.
___
*[google/cloud/pubsub/internal/streaming_subscription_batch_source.cc, line 210 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5510#-MMJDI1C9CZi7ZIIuam5:-MMJDI1D3ecJ1VFl75MZ:bjhkdp9) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4b5e820c249ec1744ae1b9a38bcee284160f13a6/google/cloud/pubsub/internal/streaming_subscription_batch_source.cc#L210)):*
> ```cpp
>         "The service was unable to fulfill your request."
>         " Please try again. [code=8a75]";
>     if (status.message() == kProbablyEmptySubscription) {
> ```

Did you get any guidance from the pubsub team about how much of this message you can/should rely upon?  If `code=8a75` is unique/stable could you match against just that? You know much more about it than I do, so maybe this is fine, just looks possibly brittle.
___
*[google/cloud/pubsub/internal/streaming_subscription_batch_source.cc, line 215 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5510#-MMJEyERDLzooyFKFSK3:-MMJEyERDLzooyFKFSK4:b-v61ms) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4b5e820c249ec1744ae1b9a38bcee284160f13a6/google/cloud/pubsub/internal/streaming_subscription_batch_source.cc#L215)):*
> ```cpp
>         shutdown_manager_->FinishedOperation("reconnect");
>       });
>       return;
> ```

I just wanted to confirm we do *not* need or want to call `shutdown_manager_->FinishedOperation("stream");` before returning here.

I'm not entirely clear on how that works but noticed `OnRetryFailure` also calls it.
___
*[google/cloud/pubsub/internal/streaming_subscription_batch_source_test.cc, line 332 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5510#-MMJGeeC7Vs6QUVH0FPJ:-MMJGeeC7Vs6QUVH0FPK:bqd3fie) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4b5e820c249ec1744ae1b9a38bcee284160f13a6/google/cloud/pubsub/internal/streaming_subscription_batch_source_test.cc#L332)):*
> ```cpp
>   uut->Shutdown();
> 
>   EXPECT_EQ(done.get(), final_status);
> ```

consider `StatusIs`

**Review by coryan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5510)* status: 1 of 2 files reviewed, 3 unresolved discussions (waiting on @devbww and @mr-salty)

___
*[google/cloud/pubsub/internal/streaming_subscription_batch_source.cc, line 201 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5510#-MMJEFwy0RuvK1HsO9tj:-MMLJC0c0_Muij7tunyg:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4b5e820c249ec1744ae1b9a38bcee284160f13a6/google/cloud/pubsub/internal/streaming_subscription_batch_source.cc#L201)):*
<details><summary><i>Previously, mr-salty (Todd Derr) wrote…</i></summary><blockquote>

> ```
>  is a retryable failure, and
> ```

I'd strike "is a retryable failure, and", it confused me - "This should be treated..." seems clear.
</blockquote></details>

Done.
___
*[google/cloud/pubsub/internal/streaming_subscription_batch_source.cc, line 210 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5510#-MMJDI1C9CZi7ZIIuam5:-MMLGz7M2xK9ds3zcOrC:b6qwh7w) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4b5e820c249ec1744ae1b9a38bcee284160f13a6/google/cloud/pubsub/internal/streaming_subscription_batch_source.cc#L210)):*
<details><summary><i>Previously, mr-salty (Todd Derr) wrote…</i></summary><blockquote>

Did you get any guidance from the pubsub team about how much of this message you can/should rely upon?  If `code=8a75` is unique/stable could you match against just that? You know much more about it than I do, so maybe this is fine, just looks possibly brittle.
</blockquote></details>

TL;DR; the error code is as hard-coded as the rest of the string.  It could change, but so could anything else in that string.

I will send you some internal links.
___
*[google/cloud/pubsub/internal/streaming_subscription_batch_source.cc, line 215 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5510#-MMJEyERDLzooyFKFSK3:-MMLHAcp9qI91YuI-9Xb:b3rp2yx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4b5e820c249ec1744ae1b9a38bcee284160f13a6/google/cloud/pubsub/internal/streaming_subscription_batch_source.cc#L215)):*
<details><summary><i>Previously, mr-salty (Todd Derr) wrote…</i></summary><blockquote>

I just wanted to confirm we do *not* need or want to call `shutdown_manager_->FinishedOperation("stream");` before returning here.

I'm not entirely clear on how that works but noticed `OnRetryFailure` also calls it.
</blockquote></details>

Yes. The purpose of using `StartOperation()` is to (as atomically as we can) check if the `shutdown_manager_` is shutting down, and if not call the body of the lambda. Once you enter that body you need to call `FinishedOperation()` (successful `StartOperation()` calls should have a matching `FinishedOperation()` call).

Candidly I do not like this whole `shutdown_manager_` idea I came up with, it has made it possible to debug some weird race conditions during shutdown, but strikes me as a crutch I built for myself.
___
*[google/cloud/pubsub/internal/streaming_subscription_batch_source_test.cc, line 332 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5510#-MMJGeeC7Vs6QUVH0FPJ:-MMLIEMtDq5HakuTiHbd:b-9fltw7) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4b5e820c249ec1744ae1b9a38bcee284160f13a6/google/cloud/pubsub/internal/streaming_subscription_batch_source_test.cc#L332)):*
<details><summary><i>Previously, mr-salty (Todd Derr) wrote…</i></summary><blockquote>

consider `StatusIs`
</blockquote></details>

I left it as-is, but I am curious why you think it would help (genuinely so, maybe I will learn something)?  `done.get()` returns a `Status`, I expect both the error code and message to be exactly the same, equality expresses that and works.
